### PR TITLE
Visualization pilots

### PR DESCRIPTION
**Pull request by zigasinko**
First visualization module - Expression Time Courses.

### CODE REVIEW

File: src/components/genexpress/common/dictyGrid/dictyGrid.tsx [hunk 1]
```diff
@@ -65,11 +69,26 @@ const DictyGrid = <T extends {}>({
         }
     };
 
+    // If any value is already selected, it needs to be manually set as selected or else grid won't show it.
+    const setSelectedData = (dataToSelect: Array<T> | undefined): void => {
+        const selectedDataIds = dataToSelect?.map((selectedDatum) => getRowId(selectedDatum));
```
Comment (dblenkus) [inline src/components/genexpress/common/dictyGrid/dictyGrid.tsx]: You can write this as `.map(getRowId)`.

It would be also nice to add this check at the beginning:
```javascript
if (!dataToSelect) return;
```
Then you can omit `?` in this line and the `if` below.

File: src/components/genexpress/common/dictyGrid/dictyGrid.tsx [hunk 2]
```diff
@@ -65,11 +69,26 @@ const DictyGrid = <T extends {}>({
...
+        if (selectedDataIds != null) {
+            gridApi.current?.forEachNode((node) => {
+                if (selectedDataIds?.includes(getRowId(node.data))) {
```
Comment (dblenkus) [inline src/components/genexpress/common/dictyGrid/dictyGrid.tsx]: Is `?` really required?

File: src/components/genexpress/common/chart/chart.tsx [hunk 3]
```diff
@@ -0,0 +1,188 @@
+import React, { ReactElement, useRef, useEffect, useCallback } from 'react';
+import * as vega from 'vega';
+import * as vegaTooltip from 'vega-tooltip';
+import { withSize, SizeMeProps } from 'react-sizeme';
+import { Spec } from 'vega';
+import { useDispatch } from 'react-redux';
+import { vegaTheme } from '../theming/vegaTheme';
+import { forwardToSentryAndNotifyUser } from '../../../../utils/errorUtils';
```
Comment (dblenkus) [inline src/components/genexpress/common/chart/chart.tsx]: `../../../../`

File: src/components/genexpress/common/chart/chart.tsx [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    dataHandlers?: Array<DataHandler>;
+    vegaSpecification: Spec;
+} & SizeMeProps;
+
+const chartPadding = { top: 40, left: 10, bottom: 10, right: 10 };
+
+/**
+ * Chart component is the basis for all charts that want to use Vega charting library.
+ *
+ * FOr performance reasons, chart will get reinitialized only if vegaSpecification changes. All other updates are handled with
```
Comment (dblenkus) [inline src/components/genexpress/common/chart/chart.tsx]: `O` -> `o`

Comment (dblenkus) [reply src/components/genexpress/common/chart/chart.tsx]: Wrap lines at a reasonable length. At most 99, but 79 would be even better.

File: src/components/genexpress/common/chart/chart.tsx [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+ * API calls (set data -> run. That's why we must pass updatableDataDefinitions for each data that can get updated.
+ *
+ * Add all your data handlers (that execute after data object changes) via 'dataHandlers' variable. All handlers will
```
Comment (dblenkus) [inline src/components/genexpress/common/chart/chart.tsx]: I think that "your" is not needed here.

File: src/components/genexpress/common/chart/chart.tsx [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+ * be automatically removed after view is removed from DOM.
+ */
+const Chart: React.FunctionComponent<ChartProps> = ({
```
Comment (dblenkus) [inline src/components/genexpress/common/chart/chart.tsx]: Why do you need `<ChartProps>` as it is already specified in L44?

File: src/components/genexpress/common/chart/chart.tsx [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+        return chartElement.current.clientHeight - chartPadding.top - chartPadding.bottom;
+    }, []);
+
+    /**
+     * Vega runAsync evaluates the underlying dataflow graph and returns a Promise that
+     * resolves upon completion of dataflow processing and scenegraph rendering.
+     */
+    const runChart = useCallback((): void => {
+        chartView.current?.runAsync().catch((error) => {
```
Comment (dblenkus) [inline src/components/genexpress/common/chart/chart.tsx]: It would be better to use `try`/`catch` as elsewhere in the codebase.

File: src/components/genexpress/modules/timeSeriesAndGeneSelector/geneSelector/selectedGenes/selectedGenes.tsx [hunk 8]
```diff
@@ -5,8 +5,8 @@ import { ContentCopy as ContentCopyIcon } from 'mdi-material-ui';
 import { Gene } from 'redux/models/internal';
 import {
     geneDeselected,
-    geneHighlighted,
     geneUnhighlighted,
+    geneHighlighted,
```
Comment (dblenkus) [inline src/components/genexpress/modules/timeSeriesAndGeneSelector/geneSelector/selectedGenes/selectedGenes.tsx:9]: Why is this moved?

File: src/components/genexpress/common/dictyGrid/dictyGrid.tsx (hunk 2 above)
Comment (zigasinko) [reply src/components/genexpress/common/dictyGrid/dictyGrid.tsx]: Yes, cause it starts out as undefined so without `?` TypeScript is complaining.

File: src/components/genexpress/modules/timeSeriesAndGeneSelector/geneSelector/selectedGenes/selectedGenes.tsx (hunk 8 above)
Comment (zigasinko) [reply src/components/genexpress/modules/timeSeriesAndGeneSelector/geneSelector/selectedGenes/selectedGenes.tsx:9]: Don't know why this was moved, probably because I've removed and added the dependency and it got sorted differently.

File: src/components/genexpress/common/chart/chart.tsx [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+            chartView.current.tooltip(tooltipHandler.call);
+
+            runChart();
+        };
+
+        generateChart();
+
+        // Vega Cleanup.
+        return (): void => {
+            // This method unregisters any timers and removes any event listeners the visualization has registered on external DOM elements.
```
Comment (dblenkus) [inline src/components/genexpress/common/chart/chart.tsx]: Wrap the line at a reasonable length.

File: src/components/genexpress/common/chart/chart.tsx [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+            // If any handlers are already attached, first remove them all.
+            addedDataHandlers.current.forEach((dataHandler) => {
+                chartView.current?.removeDataListener(dataHandler.name, dataHandler.handler);
+            });
+
+            // Add fresh handlers.
+            dataHandlers.forEach((dataHandler) => {
+                chartView.current?.addDataListener(dataHandler.name, dataHandler.handler);
+            });
+            addedDataHandlers.current = dataHandlers;
```
Comment (dblenkus) [inline src/components/genexpress/common/chart/chart.tsx:171]: Data handlers are added only if `chartView.current` is not null, is it a problem to save them here if they are not actually added?
I suggest updating the `if` above to `if (dataHandlers != null && !!chartView.current)`.

**Review by dblenkus — APPROVED** [review]
